homeassistant/components/media_player/directv.py

```diff
@@ -47,73 +55,287 @@
+        hass.data.setdefault(DATA_DIRECTV, set()).add((
+            directv_entity[CONF_HOST], directv_entity[CONF_DEVICE]))
+
+        hass.data.setdefault(DATA_DIRECTV, set()).add((
+            directv_entity[RECEIVER_ID], directv_entity[CONF_DEVICE]))
+
+        add_entities([DirecTvDevice(
+            directv_entity[CONF_NAME], directv_entity[CONF_HOST],
+            directv_entity[CONF_PORT], directv_entity[CONF_DEVICE])])
+
+        # Only enable client discovery if not disabled in configuration
+        # and this is the main DVR (not a client device itself).
+        if directv_entity[CONF_DISCOVER_CLIENTS] and \
+           directv_entity[CONF_DEVICE] == DEFAULT_DEVICE:
+            # Client discovery is to be enabled for this DVR.
+            DirecTvClientDiscovery(
+                hass, add_entities, discovered,
+                directv_entity[CONF_HOST], directv_entity[CONF_NAME],
+                directv_entity[RECEIVER_ID], directv_entity[CONF_PORT],
+                directv_entity[CONF_CLIENT_DISCOVER_INTERVAL]
+            )
+
+
+class DirecTvClientDiscovery():
+    """Discover client devices attached to DVR."""
+
+    def __init__(self, hass, add_entities, discovered, host, name,
+                 receiver_id, port=DEFAULT_PORT,
+                 interval=DEFAULT_CLIENT_DISCOVER_INTERVAL):
+        """Initialize discovery for client devices."""
+        self._hass = hass
+        self._add_entities = add_entities
+        self._discovered = discovered
+        self._host = host
+        self._name = name if name else host
+        self._port = port
+        self._receiver_id = receiver_id if receiver_id != host else None
+
+        self.dtv = None
+
+        # Client discovery to be started once HASS is started to ensure
+        # all configured devices have been added first.
+        def client_discovery_startup(event):
+            # Perform a discovery if the main entity was discovered as well or
+            # HASS started.
+            if self._discovered or event:
+                self._discover_directv_client_devices()
+
+            # Schedule discovery to run based on interval.
+            track_time_interval(
+                self._hass, self._discover_directv_client_devices,
+                interval)
+            _LOGGER.debug("%s: Client discovery scheduled for every %s",
+                          self._name, interval)
+
+        # If HASS is already running then start the discovery.
+        # If HASS is not yet running, register for the event before starting
+        # the discovery.
+        if self._hass.is_running:
+            client_discovery_startup(None)
+        else:
+            self._hass.bus.listen_once(
+                EVENT_HOMEASSISTANT_START, client_discovery_startup)
+
+    def _discover_directv_client_devices(self, now=None):
+        """Discover new client devices connected to the main DVR."""
+        known_devices = self._hass.data.get(DATA_DIRECTV, set())
```

Comment:
Just call `setdefault` once at the start of setup platform and you don't have to do this anymore.

Reply:
And I was already using setdefault :-)